internal/restorer/restorer_test.go

```diff
@@ -682,10 +693,116 @@ func TestRestorerTraverseTree(t *testing.T) {
 			// make sure we're creating a new subdir of the tempdir
 			target := filepath.Join(tempdir, "target")
 
-			err = res.traverseTree(ctx, target, string(filepath.Separator), *sn.Tree, test.Visitor(t))
+			_, err = res.traverseTree(ctx, target, string(filepath.Separator), *sn.Tree, test.Visitor(t))
 			if err != nil {
 				t.Fatal(err)
 			}
 		})
 	}
 }
+
+func normalizeFileMode(mode os.FileMode) os.FileMode {
+	if runtime.GOOS == "windows" {
+		if mode.IsDir() {
+			return 0555 | os.ModeDir
+		}
+		return os.FileMode(0444)
+	}
+	return mode
+}
+
+func checkConsistentInfo(t testing.TB, file string, fi os.FileInfo, modtime time.Time, mode os.FileMode) {
+	if fi.Mode() != mode {
+		t.Errorf("checking %q, Mode() returned wrong value, want 0%o, got 0%o", file, mode, fi.Mode())
+	}
+
+	if !fi.ModTime().Equal(modtime) {
+		t.Errorf("checking %s, ModTime() returned wrong value, want %v, got %v", file, modtime, fi.ModTime())
+	}
+}
+
+// test inspired from test case https://github.com/restic/restic/issues/1212
+func TestRestorerConsistentTimestampsAndPermissions(t *testing.T) {
+	timeForTest := time.Date(2019, time.January, 9, 1, 46, 40, 0, time.UTC)
+
+	repo, cleanup := repository.TestRepository(t)
+	defer cleanup()
+
+	_, id := saveSnapshot(t, repo, Snapshot{
+		Nodes: map[string]Node{
+			"dir": Dir{
+				Mode:    normalizeFileMode(0750 | os.ModeDir),
+				ModTime: timeForTest,
+				Nodes: map[string]Node{
+					"file1": File{
+						Mode:    normalizeFileMode(os.FileMode(0700)),
+						ModTime: timeForTest,
+						Data:    "content: file\n",
+					},
+					"anotherfile": File{
+						Data: "content: file\n",
+					},
+					"subdir": Dir{
+						Mode:    normalizeFileMode(0700 | os.ModeDir),
+						ModTime: timeForTest,
+						Nodes: map[string]Node{
+							"file2": File{
+								Mode:    normalizeFileMode(os.FileMode(0666)),
+								ModTime: timeForTest,
+								Links:   2,
+								Inode:   1,
+							},
+						},
+					},
+				},
+			},
+		},
+	})
+
+	res, err := NewRestorer(repo, id)
+	rtest.OK(t, err)
+
+	res.SelectFilter = func(item string, dstpath string, node *restic.Node) (selectedForRestore bool, childMayBeSelected bool) {
+		switch filepath.ToSlash(item) {
+		case "/dir":
+			childMayBeSelected = true
+		case "/dir/file1":
+			selectedForRestore = true
+			childMayBeSelected = false
+		case "/dir/subdir":
+			selectedForRestore = true
+			childMayBeSelected = true
+		case "/dir/subdir/file2":
+			selectedForRestore = true
+			childMayBeSelected = false
+		}
+		return selectedForRestore, childMayBeSelected
+	}
+
+	tempdir, cleanup := rtest.TempDir(t)
+	defer cleanup()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	err = res.RestoreTo(ctx, tempdir)
+	rtest.OK(t, err)
+
+	var testPatterns = []struct {
+		path    string
+		modtime time.Time
+		mode    os.FileMode
+	}{
+		{"dir", timeForTest, normalizeFileMode(0750 | os.ModeDir)},
+		{"dir/file1", timeForTest, normalizeFileMode(os.FileMode(0700))},
```

Comment:
I had in mind to use `filepath.Join("dir", "file1")` here instead of `dir/file1`. Currently the `filepath.ToSlash` call later on has no effect.

Reply:
done